luqum/utils.py

```diff
@@ -80,6 +80,128 @@ def __call__(self, tree):
         return self.visit(tree)
 
 
+class OpenRangeTransformer(visitor.TreeTransformer):
+    """Transforms open ranges to normal Range objects, i.e.
+
+    ::
+
+        >=foo            ->   [foo TO *]
+        <bar             ->   [* TO bar}
+
+    When *merge_ranges* is set (the default), this also merges open ranges in AND clauses::
+
+        >foo AND <=bar              ->   {foo TO bar]
+        [foo TO *] AND [* TO bar]   ->   [foo TO bar]
+
+    The merging of open ranges is performed by collecting all open ranges, and merging any
+    open range into any previously collected open range with an open bound. In other words, any
+    open bounds are always merged into the first suitable. Additionally, matching is always done
+    from left-to-right, so that this holds::
+
+        [a TO *] AND [b TO *] AND [* TO y] AND [* TO z]   ->   [a TO y] AND [b TO z]
```

Comment:
I have added more comments and made it a bit clearer. I have also split out one method so that the focus is on the algorithm itself. I'm unsure which case is better.